cleverhans/attacks.py

```diff
@@ -1,4 +1,374 @@
+    infinity norm (and is known as the "Fast Gradient Sign Method"). This
+    implementation extends the attack to other norms, and is therefore called
+    the Fast Gradient Method.
+    Paper link: https://arxiv.org/abs/1412.6572
+    """
+    def __init__(self, model, back='tf', sess=None, params={'eps': 0.3,
+                                                            'ord': 'np.inf',
+                                                            'y': None,
+                                                            'clip_min': None,
+                                                            'clip_max': None}):
+        """
+        Create a FastGradientMethod instance.
+
+        Attack-specific parameters:
+        :param eps: (required float) attack step size (input variation)
+        :param ord: (optional) Order of the norm (mimics Numpy).
+                    Possible values: np.inf, 1 or 2.
+        :param y: (optional) A placeholder for the model labels. Only provide
+                  this parameter if you'd like to use true labels when crafting
+                  adversarial samples. Otherwise, model predictions are used as
+                  labels to avoid the "label leaking" effect (explained in this
+                  paper: https://arxiv.org/abs/1611.01236). Default is None.
+                  Labels should be one-hot-encoded.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        super(FastGradientMethod, self).__init__(model, back, sess, params)
+
+        # Check that all required attack specific parameters are defined
+        assert 'eps' in params
+
+        # Save attack-specific parameters
+        self.eps = params['eps']
+        self.ord = params['ord'] if 'ord' in params else np.inf
+        self.y = params['y'] if 'y' in params else None
+        self.clip_min = params['clip_min'] if 'clip_min' in params else None
+        self.clip_max = params['clip_max'] if 'clip_max' in params else None
+
+        # Check if order of the norm is acceptable given current implementation
+        if self.ord not in [np.inf, int(1), int(2)]:
+            raise Exception("Norm order must be either np.inf, 1, or 2.")
+        if back == 'th' and self.ord != np.inf:
+            raise NotImplementedError("The only FastGradientMethod norm "
+                                      "implemented for Theano is np.inf.")
+
+    def generate(self, x, params={}):
+        """
+        Generate symbolic graph for adversarial examples and return.
+        """
+        if self.back == 'tf':
+            from .attacks_tf import fgm
+        else:
+            from .attacks_th import fgm
+
+        return fgm(x, self.model(x), y=self.y, eps=self.eps, ord=self.ord,
+                   clip_min=self.clip_min, clip_max=self.clip_max)
+
+    def generate_np(self, X, params={'Y': None}):
```

Comment:
The reason for this is to distinguish the placeholder from the numpy array (for instance the JSMA currently needs both)

should we switch to `x_val` and `y_val`?